### PR TITLE
State checkout has a `--set-cache` flag

### PR DESCRIPTION
**Pull request by MDrakos**
<table>
<td>
  <a href="https://activestatef.atlassian.net/browse/DX-1486" title="DX-1486" target="_blank">StoryDX-1486</a>  State checkout takes a flag that specifies the runtime dir
  </td></table>
  <br />

### CODE REVIEW

File: cmd/state/internal/cmdtree/checkout.go [hunk 1]
```diff
@@ -24,6 +24,11 @@ func newCheckoutCommand(prime *primer.Values) *captain.Command {
 				Description: locale.Tl("flag_state_checkout_branch_description", "Defines the branch to checkout"),
 				Value:       &params.Branch,
 			},
+			{
+				Name:        locale.Tl("flag_state_checkout_cache", "set-cache"),
+				Description: locale.Tl("flag_state_checkout_cache_description", "Defines the cache to use"),
```
Comment (mitchell-as) [inline cmd/state/internal/cmdtree/checkout.go:29]: This seems pretty vague. Is this the language we want to use?

File: pkg/projectfile/projectfile.go [hunk 2]
```diff
@@ -972,9 +975,51 @@ func createCustom(params *CreateParams, lang language.Language) (*Project, error
...
+	}
+
+	fileContents, err := strutils.ParseTemplate(string(tplContents), data)
+	if err != nil {
+		return errs.Wrap(err, "Could not parse %s", tplName)
+	}
+
+	// Username on Windows is  DOMAIN\username, we only want the username
+	// rather than trimming the domain, we just take the last part of the
+	// home directory's path
```
Comment (mitchell-as) [inline pkg/projectfile/projectfile.go:1011]: Some users have different usernames than their home directory (not just on Windows, but on any platform). Some usernames may also have spaces in them. Is this going to cause any trouble with the logic here?

**Review by mitchell-as — APPROVED** [review]
This looks fine in principle to me, but I made some comments worth considering.

File: pkg/projectfile/projectfile.go (hunk 2 above)
Comment (MDrakos) [reply pkg/projectfile/projectfile.go]: Good point. Updated the username handling for windows and also ensure that the username has no spaces on any platforms.

**Review by mitchell-as — DISMISSED** [review]

File: pkg/projectfile/projectfile.go [hunk 3]
```diff
@@ -972,9 +975,46 @@ func createCustom(params *CreateParams, lang language.Language) (*Project, error
...
+		return errs.Wrap(err, "Could not read asset")
+	}
+
+	fileContents, err := strutils.ParseTemplate(string(tplContents), data)
+	if err != nil {
+		return errs.Wrap(err, "Could not parse %s", tplName)
+	}
+
+	// Trim any non-alphanumeric characters from the username
+	nonAlphanumericRegex := regexp.MustCompile(`[^a-zA-Z0-9 ]+`)
```
Comment (Naatan) [inline pkg/projectfile/projectfile.go:1010]: Please put this at the top in the root of the scope.

File: pkg/projectfile/projectfile.go [hunk 4]
```diff
@@ -972,9 +975,46 @@ func createCustom(params *CreateParams, lang language.Language) (*Project, error
 		return nil, err
 	}
 
+	if params.Cache != "" {
+		createErr := createCacheFile(params.Directory, params.Cache)
```
Comment (Naatan) [inline pkg/projectfile/projectfile.go:979]: Can we name this `createHostFile()`? Cache is just the first usage of it. But I foresee we may want to use this for other things in the future.

File: internal/assets/contents/activestate.yaml.cache.tpl [hunk 5]
```diff
@@ -0,0 +1 @@
+cache: {{ .Cache }}
```
Comment (Naatan) [inline internal/assets/contents/activestate.yaml.cache.tpl:1]:
```suggestion
# It is recommended that you do not commit this file as it contains configuration that is specific to your machine
cache: {{ .Cache }}
```

**Review by Naatan — CHANGES_REQUESTED** [review]

File: internal/assets/contents/activestate.yaml.cache.tpl [hunk 6]
```diff
@@ -0,0 +1,2 @@
+# It is recommended that you do not commit this file as it contains configuration that is specific to your machine
+cache: {{ .Cache }}
```
Comment (Naatan) [inline internal/assets/contents/activestate.yaml.cache.tpl:2]:
```suggestion
cache: {{ .Cache }}

```

**Review by Naatan — CHANGES_REQUESTED** [review]
Looks good! Really a minor change just to add a line break at the end. Normally not something I'd nitpick about but that file goes out to users.

**Review by Naatan — APPROVED** [review]

**Pull request merged**